### ISSUE TITLE
Nested SliverStickyHeader

### ISSUE THREAD

**Issue by bunopus** [letsar/flutter_sticky_header#5]
Thank you for your library, it's super useful. I have one question: Is it possible to achieve something like this:

<img src="https://user-images.githubusercontent.com/9258200/42099203-7cd66346-7bc5-11e8-8914-850639d01784.gif" width="400px">  

As you can see header with dates sticks to the screen as well as the side header with time (it is a https://github.com/google/iosched app). Example app have great section with side headers, but i tried to nest `SliverStickyHeader` and it doesn't work. Second Header doesn't stick to screen.

<img src="https://user-images.githubusercontent.com/9258200/42100218-483d02b8-7bc8-11e8-834c-2afab5b89728.gif" width="400px">

Code
```dart
// code from example app
class MainScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    List<Widget> slivers = new List<Widget>();

    // App Bar
    slivers.add(getAppBar());

    slivers.add( new SliverStickyHeader(
      // Tabs header
      header:_buildHeader(1, text: 'Tabs'),
      // Side header
      sliver: SliverStickyHeader(
       overlapsContent: true,
       header: _buildSideHeader(1),
       sliver: new SliverPadding(
       // other code is same, create grid

  }

  Widget getAppBar() {
    return SliverAppBar(
      expandedHeight: 60.0,
      floating: true,
      brightness: Brightness.light,
      title: Text('Main AppBar',
          style: TextStyle(color: Colors.black.withOpacity(1.0))),
      elevation: 0.0,
      centerTitle: true,
      backgroundColor: Color.fromRGBO(255, 255, 255, 1.0),
    );
  }
```

**Comment by letsar** [comment]
Hi @bunopus, thanks for the feedback 😃.
All the slivers are in a CustomScrollView I presume?
I am working on a way to have Nested SliverStickyHeader for another feature, but for now I am stucked 😞 .
I will have another look to this thanks to your example.
I will keep you posted.

**Comment by letsar** [comment]
I reproduced the issue, and saw that the `SliverStickyHeader` was painted under the AppBar. I think I will have to play with the `SliverOverlapAbsorber` to fix this.

![sticky_header_pinned_under_app_bar](https://user-images.githubusercontent.com/9378033/42123667-2eacd272-7c56-11e8-8ed5-217345a60373.gif)

**Comment by letsar** [comment]
I found how to do it :grin:!

![sticky_header_pinned_app_bar_ok](https://user-images.githubusercontent.com/9378033/42125213-a4a17548-7c72-11e8-8379-be3c6ce1a045.gif)

It's a little bit complex since you'll have to play with `SliverOverlapAbsorber`. If you find a simpler solution will you share it here please?

You can find the entire code for the view above here: https://gist.github.com/letsar/2e3cc98d328b3e84170abacf154e545f

**Comment by bunopus** [comment]
@letsar Wow! 😍
Thank you so much! Will try it right now!

**Comment by letsar** [comment]
You're welcome 😄 .
If you find this library useful, don't hesitate to ⭐️ it 😉

**Comment by bunopus** [comment]
Tried your solution, and noticed one thing: app bar title ("Main AppBar") remains on the screen with tab bar. It's not exactly what i want. I think user don't need to see title all the time. In the original Google IO app (in first post) app bar title hides when user scrolls page. Maybe you know how to achieve that? Found PR https://github.com/flutter/flutter/pull/8345 and it says that you can pass `floating: true, pinned: true` but it doesn't work :-(

**Comment by letsar** [comment]
Your're welcome and yes you're right it's not exactly what you wanted but it's more a `SliverAppBar `issue than a `SliverStickyHeader `one :wink:.

**Comment by workerbee22** [comment]
There is a simpler solution to use with a **TabBar** that avoids the complicated use of **SliverOverlapAbsorber**

@bunopus you'll notice it is very much like the Google IO app layout and tab/scrolling behaviour you want 👍

```dart
import 'package:flutter/material.dart';
import 'package:flutter_sticky_header/flutter_sticky_header.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Sliver AppBar with TabBar and Sticky',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Sliver AppBar with TabBar and Sticky'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

// Note - SingleTickerProviderStateMixin, see TabController class API page sample
class _MyHomePageState extends State<MyHomePage>
    with SingleTickerProviderStateMixin {
  // Instantiate TabController and ScrollViewController
  TabController _tabController;
  ScrollController _scrollViewController;

  @override
  void initState() {
    super.initState();
    _tabController = TabController(vsync: this, length: 4);
    _scrollViewController = ScrollController();
  }

  @override
  void dispose() {
    _tabController.dispose();
    _scrollViewController.dispose();
    super.dispose();
  }

  final count = 30;
  final sliverIndex = 0;
  final firstIndex = 0;

  Widget _buildSideHeader(BuildContext context, int index, {String text}) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 8.0, vertical: 4.0),
      child: Align(
        alignment: Alignment.centerLeft,
        child: new SizedBox(
          height: 44.0,
          width: 44.0,
          child: GestureDetector(
            onTap: () => Scaffold.of(context)
                .showSnackBar(new SnackBar(content: Text('$index'))),
            child: new CircleAvatar(
              backgroundColor: Colors.redAccent,
              foregroundColor: Colors.white,
              child: new Text('$index'),
            ),
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: NestedScrollView(
        controller: _scrollViewController,
        headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
          // These are the slivers that show up in the "outer" scroll view.
          return <Widget>[
            SliverAppBar(
              title: Text('Schedule'), // This is the title in the app bar.
              centerTitle: true,
              // pinned applies to whole AppBar including TabBar
              // but TabBar always stays on screen for some reason
              pinned: true,
              floating: true,
              forceElevated: innerBoxIsScrolled,
              leading: Icon(Icons.person),
              bottom: TabBar(
                tabs: <Widget>[
                  Tab(
                    text: 'Day 1',
                  ),
                  Tab(
                    text: 'Day 2',
                  ),
                  Tab(
                    text: 'Day 3',
                  ),
                  Tab(
                    text: 'Agenda',
                  ),
                ],
                controller: _tabController,
              ),
            ),
          ];
        },
        body: TabBarView(
          // These are the contents of the tab views, below the tabs.
          children: <Widget>[
            CustomScrollView(
              controller: _scrollViewController,
              slivers: List.generate(count, (sliverIndex) {
                sliverIndex += firstIndex;
                // *** SliverStickyHeader - header is static and built by _BuildSideHeader
                return new SliverStickyHeader(
                  overlapsContent: true,
                  header: _buildSideHeader(context, sliverIndex),
                  sliver: new SliverPadding(
                    // Padding on LEFT to make room for overlapping side header
                    padding: new EdgeInsets.only(left: 60.0),
                    // SliverGrid of 3 tiles
                    sliver: new SliverGrid(
                      // Grid tiles are spaced by 4.0 all around, there is 3 and all square
                      gridDelegate:
                          new SliverGridDelegateWithFixedCrossAxisCount(
                              crossAxisCount: 3,
                              crossAxisSpacing: 4.0,
                              mainAxisSpacing: 4.0,
                              childAspectRatio: 1.0),
                      // Grid tiles x 12 have touch to show snackbar, orange and footer text
                      delegate: new SliverChildBuilderDelegate(
                        (context, i) => GestureDetector(
                              onTap: () => Scaffold.of(context).showSnackBar(
                                  new SnackBar(content: Text('Grid tile #$i'))),
                              child: new GridTile(
                                child: Card(
                                  child: new Container(
                                    color: Colors.orange,
                                  ),
                                ),
                                footer: new Container(
                                  color: Colors.white.withOpacity(0.5),
                                  child: Padding(
                                    padding: const EdgeInsets.all(8.0),
                                    child: new Text(
                                      'Grid tile #$i',
                                      style:
                                          const TextStyle(color: Colors.black),
                                    ),
                                  ),
                                ),
                              ),
                            ),
                        childCount: 6,
                      ),
                    ),
                  ),
                );
              }),
            ),
            Center(
              child: Text('Day 2 Tab'),
            ),
            Center(
              child: Text('Day 3 Tab'),
            ),
            Center(
              child: Text('Agenda Tab'),
            ),
          ],
          controller: _tabController,
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.filter_list),
        onPressed: () {},
      ),
    );
  }
}

```

**Comment by workerbee22** [comment]
@bunopus You said you achieved this  by adding another SliverAppBar but this means you have 2 SliverAppBars? Your gist link code gives me 2 x bars. 

https://gist.github.com/bunopus/4a99dc9def2932e0aff2629f3905093a#file-tab_bar_view-dart-L36-L39

How did you make sure the second SliverAppBar ONLY has the TabBar in it?

**Comment by bunopus** [comment]
@workerbee22 
> How did you make sure the second SliverAppBar ONLY has the TabBar in it?
Actually it turned out that it leads to bugs, so i leave it with Tab Bar on the screen, as Romain showed few messages above.

**Comment by cgestes** [comment]
This fork has working nested SliverStickyHeaders 🗡️ 

https://github.com/UnderKoen/flutter_sticky_header

**Comment by fennel-ptorchia** [comment]
Is there an updated solution for this issue? This does not appear to work when using tabs and the nested scroll view, I'm looking to be able to pin headers that are inside of a custom scroll view in a tab bar view in the body of a nested scroll view.

**Comment by PixelPatrik** [comment]
When testing, my application crashed while changing tabs, because the handle did not remove listeners correctly. I needed to add the 'detach' function and remove the listener accordingly.
```
  @override
  void detach() {
    handle.removeListener(markNeedsLayout);
    super.detach();
  }
  ```
  
  Here are the snippets if anyone is interested.
  
  ```
/// A sliver that has a sliver geometry based on the values stored in a
/// [SliverOverlapAbsorberHandle].
///
/// The [RenderSliverOverlapAbsorber] must be an earlier descendant of a common
/// ancestor [RenderViewport] (probably a [RenderNestedScrollViewViewport]), so
/// that it will always be laid out before the [RenderSliverObstructionInjector]
/// during a particular frame.
class RenderSliverObstructionInjector extends RenderSliverOverlapInjector {
  /// Creates a sliver that is as tall as the value of the given [handle]'s extent.
  ///
  /// The [handle] must not be null.
  RenderSliverObstructionInjector({
    required SliverOverlapAbsorberHandle handle,
    RenderSliver? child,
  })  : _handle = handle,
        super(handle: handle);

  double _currentLayoutExtent = 0;
  double _currentMaxExtent = 0;

  /// The object that specifies how wide to make the gap injected by this render
  /// object.
  ///
  /// This should be a handle owned by a [RenderSliverOverlapAbsorber] and a
  /// [RenderNestedScrollViewViewport].
  @override
  SliverOverlapAbsorberHandle get handle => _handle;
  SliverOverlapAbsorberHandle _handle;
  @override
  set handle(SliverOverlapAbsorberHandle value) {
    if (handle == value) return;
    if (attached) {
      handle.removeListener(markNeedsLayout);
    }
    _handle = value;
    if (attached) {
      handle.addListener(markNeedsLayout);
      if (handle.layoutExtent != _currentLayoutExtent ||
          handle.scrollExtent != _currentMaxExtent) markNeedsLayout();
    }
  }

  @override
  void attach(PipelineOwner owner) {
    super.attach(owner);
    handle.addListener(markNeedsLayout);
    if (handle.layoutExtent != _currentLayoutExtent ||
        handle.scrollExtent != _currentMaxExtent) {
      markNeedsLayout();
    }
  }

  @override
  void detach() {
    handle.removeListener(markNeedsLayout);
    super.detach();
  }

  @override
  void performLayout() {
    _currentLayoutExtent = handle.layoutExtent ?? 0;
    _currentMaxExtent = handle.layoutExtent ?? 0;
    geometry = SliverGeometry(
      scrollExtent: 0.0,
      paintExtent: _currentLayoutExtent,
      maxPaintExtent: _currentMaxExtent,
    );
  }
}
```

```
class SliverObstructionInjector extends SliverOverlapInjector {
  /// Creates a sliver that is as tall as the value of the given [handle]'s
  /// layout extent.
  ///
  /// The [handle] must not be null.
  ///
  const SliverObstructionInjector(
      {super.key, required SliverOverlapAbsorberHandle handle, super.sliver})
      : super(handle: handle);

  @override
  RenderSliverObstructionInjector createRenderObject(BuildContext context) {
    return RenderSliverObstructionInjector(
      handle: handle,
    );
  }
}
```
